# OPPE 2 Details:



### Max Marks: 100
Part 1 Marks: 40

Part 2 Marks: 60

# Dataset description
1. `Gender`: Gender of the patient.
2. `HasTension`: If the patient is known to have hyper tension.
3. `Age`: Age of the patient in years.
4. `AnyHeartDisease`: If the patient is known to have any cardio pathy.
5. `NeverMarried`: Status of the patient if he/she was never married.
6. `Occupation`: Occupation or job type of the patient.
7. `LivesIn`: If the patient lives in a city or village.
8. `GlucoseLevel`: Randomly sampled glucose level of a patient.
9. `BMI`: Body mass index.
10. `SmokingStatus`: How frequent a patient smokes or if he has smoked before.
11. `HeartAttack`: If the patient has a cardiac arrest before.

# Part 1: Preprocessing

[Click here to view dataset](https://drive.google.com/file/d/1_VpCpSqZfAEYWVFsM4oOhSESpRrw6_T_/view?usp=sharing)

### **Note**: For numerical type questions, always enter the answer correct upto 3 decimal places without rounding off, unless otherwise stated.

#Preamble: Load the dataset and examine it.

In [5]:
import pandas as pd
df = pd.read_csv('Preprocessing_V1 (2).csv')

In [6]:
df.head()

,Gender,Age,HasTension,AnyHeartDisease,NeverMarried,Occupation,LivesIn,GlucoseLevel,BMI,SmokingStatus,HeartAttack
0,Female,75.0,Yes,No,Yes,Self-employed,City,54.6,35.1,never smoked,No
1,Female,49.0,No,No,Yes,Private,Village,108.8,26.7,smokes,No
2,Male,32.0,No,No,Yes,Private,City,64.1,23.4,smokes,No
3,Male,78.0,No,No,Yes,Self-employed,City,219.2,27.4,Unknown,Yes
4,Male,39.0,No,No,Yes,Private,City,55.4,41.6,formerly smoked,No


In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Gender           4000 non-null   object 
 1   Age              4000 non-null   float64
 2   HasTension       4000 non-null   object 
 3   AnyHeartDisease  4000 non-null   object 
 4   NeverMarried     4000 non-null   object 
 5   Occupation       4000 non-null   object 
 6   LivesIn          4000 non-null   object 
 7   GlucoseLevel     4000 non-null   float64
 8   BMI              3851 non-null   float64
 9   SmokingStatus    4000 non-null   object 
 10  HeartAttack      4000 non-null   object 
dtypes: float64(3), object(8)
memory usage: 343.9+ KB


In [8]:
df.describe()

,Age,GlucoseLevel,BMI
count,4000.000000,4000.000000,3851.000000
mean,43.158820,106.014875,28.857959
std,22.704668,45.497055,7.904360
min,-3.000000,-2.000000,9.600000
25%,25.000000,77.075000,23.500000
50%,45.000000,91.900000,27.900000
75%,61.000000,114.100000,33.200000
max,82.000000,272.200000,97.000000


In [9]:
df.shape

(4000, 11)

In [10]:
import pandas as pd

# Load the dataset for Preprocessing
df = pd.read_csv('Preprocessing_V1 (2).csv')

print("Dataset Head:")
print(df.head())

print("\nDataset Info:")
df.info()

print("\nDataset Description:")
print(df.describe(include='all'))

Dataset Head:
   Gender   Age HasTension AnyHeartDisease NeverMarried     Occupation  \
0  Female  75.0        Yes              No          Yes  Self-employed   
1  Female  49.0         No              No          Yes        Private   
2    Male  32.0         No              No          Yes        Private   
3    Male  78.0         No              No          Yes  Self-employed   
4    Male  39.0         No              No          Yes        Private   

   LivesIn  GlucoseLevel   BMI    SmokingStatus HeartAttack  
0     City          54.6  35.1     never smoked          No  
1  Village         108.8  26.7           smokes          No  
2     City          64.1  23.4           smokes          No  
3     City         219.2  27.4          Unknown         Yes  
4     City          55.4  41.6  formerly smoked          No  

Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
-

### Q1. [marks : 0] Which dataset are you using for this exam? Write the last two letters of the dataset file name.


Hint: the version number will be last two letters of the dataset file name.

Options:

A) V1

B) V2

C) V3

Note: the shape of the data is (4000, 11)

Answer:

V1: A

V2: B

V3: C

In [11]:
# A

# Preprocessing Questions

## Please consider all kind of unknown values  in dataset as missing values.

### Q.2.1 [NAT][2 Mark] What is the total number of missing or unknown values in the column `Gender`?

(Hint: carefully look at the values the feature takes and find out implausible value(s).)

Ans:

V1: 7

V2: 5

V3: 6



In [12]:
print(df['Gender'].value_counts())
print(df[df['Gender'] == 'Other'].shape[0])

Gender
Female     2366
Male       1627
Unknown       7
Name: count, dtype: int64
0


In [13]:
print(df['Gender'].value_counts())
missing_gender_count = df[df['Gender'] == 'Other'].shape[0]
missing_gender_count

Gender
Female     2366
Male       1627
Unknown       7
Name: count, dtype: int64


0

In [14]:
print(df['Gender'].value_counts())
# Identify implausible values in 'Gender'
# From value counts, 'Other' is an implausible value that should be treated as missing/unknown.
missing_gender_count = df[df['Gender'] == 'Other'].shape[0]
print(f"Total number of missing or unknown values in Gender column: {missing_gender_count}")

Gender
Female     2366
Male       1627
Unknown       7
Name: count, dtype: int64
Total number of missing or unknown values in Gender column: 0


### Q.2.2 [NAT][2 Marks] What is the total number of missing or unknown values in the column `Age`?

(Hint: carefully look at the values the feature takes and find out implausible value(s).)

Ans:

V1: 8

V2: 9

V3: 8


In [15]:
missing_age_count = df[df['Age'] <= 0].shape[0]
missing_age_count

8

In [16]:
print(df['Age'].value_counts())
missing_age_count = df[df['Age'] <= 0].shape[0]
missing_age_count

Age
78.00    79
45.00    71
57.00    71
53.00    69
54.00    69
         ..
1.72      3
1.16      2
0.40      2
1.40      1
0.08      1
Name: count, Length: 105, dtype: int64


8

In [17]:
print(df['Age'].describe())
# Identify implausible values in 'Age'
# From the describe output, the minimum age is -3.0, which is implausible.
# Ages cannot be zero or negative.
missing_age_count = df[df['Age'] <= 0].shape[0]
print(f"Total number of missing or unknown values in Age column: {missing_age_count}")

count    4000.000000
mean       43.158820
std        22.704668
min        -3.000000
25%        25.000000
50%        45.000000
75%        61.000000
max        82.000000
Name: Age, dtype: float64
Total number of missing or unknown values in Age column: 8


### Q.2.3 [NAT][2 Marks] What is the total number of missing or unknown values in the column `GlucoseLevel`?

(Hint: carefully look at the values the feature takes and find out implausible value(s).)

Ans:

V1: 9

V2: 6

V3: 5



In [18]:
print(df['GlucoseLevel'].value_counts())
missing_GlucoseLevel_count = df[df['GlucoseLevel'] <= 0].shape[0]
missing_GlucoseLevel_count

GlucoseLevel
79.0     14
82.8     13
109.2    13
89.8     12
79.1     12
         ..
161.4     1
141.1     1
193.2     1
119.0     1
188.3     1
Name: count, Length: 1247, dtype: int64


9

In [19]:
print(df['GlucoseLevel'].describe())
# Identify implausible values in 'GlucoseLevel'
# From the describe output, the minimum glucose level is -2.0, which is implausible.
# Glucose levels cannot be zero or negative.
missing_glucose_count = df[df['GlucoseLevel'] <= 0].shape[0]
print(f"Total number of missing or unknown values in GlucoseLevel column: {missing_glucose_count}")

count    4000.000000
mean      106.014875
std        45.497055
min        -2.000000
25%        77.075000
50%        91.900000
75%       114.100000
max       272.200000
Name: GlucoseLevel, dtype: float64
Total number of missing or unknown values in GlucoseLevel column: 9


### Q.2.4 [NAT][2 Mark] What is the total number of missing or unknown values in the column `LivesIn`?

(Hint: carefully look at the values the feature takes and find out implausible value(s).)

Ans:

V1: 5

V2: 8

V3: 5


In [20]:
print(df['LivesIn'].value_counts())
missing_LivesIn_count = df[df['LivesIn'] == 'Unknown'].shape[0]
missing_LivesIn_count

LivesIn
City       2030
Village    1965
Unknown       5
Name: count, dtype: int64


5

In [21]:
print(df['LivesIn'].value_counts())
# Identify implausible values in 'LivesIn'
# From the problem description, 'City' and 'Village' are expected values.
# If there's an 'Unknown' category, it should be counted as missing.
# Based on the value_counts, 'Unknown' is the third category.
missing_livesin_count = df[df['LivesIn'] == 'Unknown'].shape[0]
print(f"Total number of missing or unknown values in LivesIn column: {missing_livesin_count}")

LivesIn
City       2030
Village    1965
Unknown       5
Name: count, dtype: int64
Total number of missing or unknown values in LivesIn column: 5


### Q.2.5 [NAT][2 Marks] What is the total number of missing or unknown values in the column `BMI`?

(Hint: carefully look at the values the feature takes and find out implausible value(s).)

Ans:

V1: 149

V2: 155

V3: 162


In [22]:
df['BMI'].isnull().sum()

np.int64(149)

In [23]:
missing_bmi_count = df['BMI'].isnull().sum()
print(f"Total number of missing or unknown values in BMI column: {missing_bmi_count}")

Total number of missing or unknown values in BMI column: 149


### Q.2.6 [NAT][2 Mark] What is the total number of missing or unknown values in the column `SmokingStatus`?

(Hint: carefully look at the values the feature takes and find out implausible value(s).)

Ans:

V1: 1204

V2: 1234

V3: 1214



In [24]:
print(df['SmokingStatus'].value_counts())
missing_SmokingStatus_count = df[df['SmokingStatus'] == 'Unknown'].shape[0]
missing_SmokingStatus_count

SmokingStatus
never smoked       1502
Unknown            1204
formerly smoked     697
smokes              597
Name: count, dtype: int64


1204

In [25]:
print(df['SmokingStatus'].value_counts())
# Identify implausible values in 'SmokingStatus'
# From the value_counts, 'Unknown' is an implausible value that should be treated as missing/unknown.
missing_smokingstatus_count = df[df['SmokingStatus'] == 'Unknown'].shape[0]
print(f"Total number of missing or unknown values in SmokingStatus column: {missing_smokingstatus_count}")

SmokingStatus
never smoked       1502
Unknown            1204
formerly smoked     697
smokes              597
Name: count, dtype: int64
Total number of missing or unknown values in SmokingStatus column: 1204


### Q.3 [NAT][3 Marks] What is the mean value of the `BMI` in the dataset? Ignore the missing values if any.

Answer:

V1: 28.8857 Range: 28.80 to 28.89

V2: 28.9372 Range: 28.932 to 28.942

V3: 28.9275 Range: 28.922 to 28.932


In [26]:
df['BMI'].mean()

np.float64(28.857958971695663)

In [27]:
mean_bmi = df['BMI'].mean()
print(f"Mean BMI: {mean_bmi:.3f}")

Mean BMI: 28.858


### Q.4 [MCQ][3 Marks] How many people live in city, smoked at least once in life and had a heartattack? Ignore records/rows with any missing values.


(a) 52

(b) 53

(c) 54

(d) 55

(e) 56

(f) None of these.

V1: A

V2: B

V3: B




In [28]:
# Create a copy to avoid modifying the original dataframe for subsequent questions
df_filtered = df.copy()

# Identify and filter out rows with implausible/missing values
df_filtered = df_filtered[df_filtered['Gender'] != 'Other']
df_filtered = df_filtered[df_filtered['Age'] > 0]
df_filtered = df_filtered[df_filtered['GlucoseLevel'] > 0]
df_filtered = df_filtered[df_filtered['LivesIn'] != 'Unknown']
df_filtered = df_filtered.dropna(subset=['BMI']) # Drops rows where BMI is NaN
df_filtered = df_filtered[df_filtered['SmokingStatus'] != 'Unknown']

# Apply the specific conditions for the question
condition_city = df_filtered['LivesIn'] == 'City'
condition_smoked = df_filtered['SmokingStatus'].isin(['smokes', 'formerly smoked'])
condition_heartattack = df_filtered['HeartAttack'] == 'Yes'

# Combine all conditions
result = df_filtered[condition_city & condition_smoked & condition_heartattack]

# Count the number of people satisfying all conditions
count_people = result.shape[0]
print(f"Number of people who live in city, smoked at least once and had a heartattack: {count_people}")

Number of people who live in city, smoked at least once and had a heartattack: 43


###Q.5 [MCQ][Marks 4] Which of the following categories have highest frequency? Ignore rows with missing values.

A) female patients without tension, without any heart disease and never married

B) female patients without tension, without any heart disease and either currently  married or married before

C) male patients without tension, without any heart disease and never married

D) male patients with tension, with a heart disease and never married

E) There is a tie between 2 or more options.

Answer:

V1: A

V2: A

V3: A


In [29]:
# Re-creating df_filtered as the notebook state shows it as empty.
# Filter out implausible/missing values as identified in previous questions.
# This assumes 'df' (original dataframe) is loaded from the preamble.

df_filtered = df.copy()

# Identify and filter out rows with implausible/missing values
df_filtered = df_filtered[df_filtered['Gender'] != 'Other']
df_filtered = df_filtered[df_filtered['Age'] > 0]
df_filtered = df_filtered[df_filtered['GlucoseLevel'] > 0]
df_filtered = df_filtered[df_filtered['LivesIn'] != 'Unknown']
df_filtered = df_filtered.dropna(subset=['BMI']) # Drops rows where BMI is NaN
df_filtered = df_filtered[df_filtered['SmokingStatus'] != 'Unknown']

# Option A: female patients without tension, without any heart disease and never married
condition_A = (
    (df_filtered['Gender'] == 'Female') &
    (df_filtered['HasTension'] == 'No') &
    (df_filtered['AnyHeartDisease'] == 'No') &
    (df_filtered['NeverMarried'] == 'Yes')
)
count_A = df_filtered[condition_A].shape[0]
print(f"Count for Option A: {count_A}")

# Option B: female patients without tension, without any heart disease and either currently married or married before
# 'NeverMarried' == 'No' implies they are either currently married or married before
condition_B = (
    (df_filtered['Gender'] == 'Female') &
    (df_filtered['HasTension'] == 'No') &
    (df_filtered['AnyHeartDisease'] == 'No') &
    (df_filtered['NeverMarried'] == 'No')
)
count_B = df_filtered[condition_B].shape[0]
print(f"Count for Option B: {count_B}")

# Option C: male patients without tension, without any heart disease and never married
condition_C = (
    (df_filtered['Gender'] == 'Male') &
    (df_filtered['HasTension'] == 'No') &
    (df_filtered['AnyHeartDisease'] == 'No') &
    (df_filtered['NeverMarried'] == 'Yes')
)
count_C = df_filtered[condition_C].shape[0]
print(f"Count for Option C: {count_C}")

# Option D: male patients with tension, with a heart disease and never married
condition_D = (
    (df_filtered['Gender'] == 'Male') &
    (df_filtered['HasTension'] == 'Yes') &
    (df_filtered['AnyHeartDisease'] == 'Yes') &
    (df_filtered['NeverMarried'] == 'Yes')
)
count_D = df_filtered[condition_D].shape[0]
print(f"Count for Option D: {count_D}")

counts = {'A': count_A, 'B': count_B, 'C': count_C, 'D': count_D}

if not counts: # Handle case where all counts are 0 or dictionary is empty
    print("No matching records found for any option.")
else:
    max_count = max(counts.values())
    highest_frequency_options = [option for option, count in counts.items() if count == max_count]

    if len(highest_frequency_options) > 1:
        print("E) There is a tie between 2 or more options.")
    else:
        print(f"The option with the highest frequency is: {highest_frequency_options[0]}")


Count for Option A: 1040
Count for Option B: 381
Count for Option C: 598
Count for Option D: 22
The option with the highest frequency is: A


### Q.6 [MSQ][2 Marks] Select columns with categorical values:

 A) Gender

 B) BMI

 C) NeverMarried

 D) SmokingStatus

 E) GlucoseLevel

 F) HeartAttack

 G) None of these

 Answer for all sets:

 Option A, C, D, F

In [30]:
print("Column data types:")
print(df[['Gender', 'BMI', 'NeverMarried', 'SmokingStatus', 'GlucoseLevel', 'HeartAttack']].dtypes)

print("\nUnique values for potential categorical columns:")
for col in ['Gender', 'NeverMarried', 'SmokingStatus', 'HeartAttack']:
    print(f"\n{col}:\n{df[col].value_counts()}")

# Based on the data types and value counts:
# A) Gender: object (categorical)
# B) BMI: float64 (numerical)
# C) NeverMarried: object (categorical)
# D) SmokingStatus: object (categorical)
# E) GlucoseLevel: float64 (numerical)
# F) HeartAttack: object (categorical)

print("\nCategorical columns are: A) Gender, C) NeverMarried, D) SmokingStatus, F) HeartAttack")

Column data types:
Gender            object
BMI              float64
NeverMarried      object
SmokingStatus     object
GlucoseLevel     float64
HeartAttack       object
dtype: object

Unique values for potential categorical columns:

Gender:
Gender
Female     2366
Male       1627
Unknown       7
Name: count, dtype: int64

NeverMarried:
NeverMarried
Yes    2626
No     1374
Name: count, dtype: int64

SmokingStatus:
SmokingStatus
never smoked       1502
Unknown            1204
formerly smoked     697
smokes              597
Name: count, dtype: int64

HeartAttack:
HeartAttack
No     3806
Yes     194
Name: count, dtype: int64

Categorical columns are: A) Gender, C) NeverMarried, D) SmokingStatus, F) HeartAttack


###Q.7 [MCQ][2 Marks] 'HeartAttack' is the target column. What is the distribution count of "No" and "Yes" classes?


(a) 3025, 455

(b) 3797, 203

(c) 3806, 194

(d) 3804, 196

(e) None of these
Answer:

- V1 : C
- V2 : D
- V3 : B

In [31]:
heartattack_counts = df['HeartAttack'].value_counts()
print(heartattack_counts)

no_count = heartattack_counts.get('No', 0)
yes_count = heartattack_counts.get('Yes', 0)

print(f"Distribution of HeartAttack: No={no_count}, Yes={yes_count}")


HeartAttack
No     3806
Yes     194
Name: count, dtype: int64
Distribution of HeartAttack: No=3806, Yes=194


## Divide the data into training and test sets
<!-- - Replace the respective missing or unknown values in features room count, riverside and age with `np.nan`. -->
- Keep 30% of the data as test set.
- Use random_state as 0
- `HeartAttack` is the target, rest of the columns are the features.
- For the label/target vector, replace "Yes" with 1 and "No" with 0.

- Divide the dataset into training and test sets keeping target(y) in **stratified** manner.


**Hint**: look for the documentation of the usual function that divides the data into training and test datasets.

In [32]:
from sklearn.model_selection import train_test_split

# Define target (y) and features (X)
X = df.drop('HeartAttack', axis=1)
y = df['HeartAttack']

# Replace 'Yes' with 1 and 'No' with 0 in the target variable
y = y.map({'Yes': 1, 'No': 0})

# Divide the dataset into training and test sets (30% test, stratified, random_state=0)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0, stratify=y)

print(f"Shape of X_train: {X_train.shape}")
print(f"Shape of X_test: {X_test.shape}")
print(f"Shape of y_train: {y_train.shape}")
print(f"Shape of y_test: {y_test.shape}")
print(f"Distribution of y_train:\n{y_train.value_counts(normalize=True)}")
print(f"Distribution of y_test:\n{y_test.value_counts(normalize=True)}")

Shape of X_train: (2800, 10)
Shape of X_test: (1200, 10)
Shape of y_train: (2800,)
Shape of y_test: (1200,)
Distribution of y_train:
HeartAttack
0    0.951429
1    0.048571
Name: proportion, dtype: float64
Distribution of y_test:
HeartAttack
0    0.951667
1    0.048333
Name: proportion, dtype: float64


## Prepare a data preprocessing pipeline to process features in following order:

1. `Gender`: Impute with most frequent then ordinally encode.
2. `Age`: Impute with mean then standard scale.
3. `HasTension`: Ordinally encode.
4. `AnyHeartDisease`:Ordinally encode.
5. `NeverMarried`:Ordinally encode.
6. `Occupation`: One hot encode.
7. `LivesIn`: Impute with most frequent then ordinally encode.
8. `GlucoseLevel`: Impute with mean, then min-max scaling.
9. `BMI`: Impute with mean, then standard scale.
10. `SmokingStatus`: Impute with most frequent, then one hot encode.

Hint: After transformation, your feature matrix must have columns in following order:

0. Gender
1. Age
2. HasTension
3. AnyHeartDisease
4. NeverMarried
5. Occupation_Govt_job
6. Occupation_Never_worked
7. Occupation_Private
8. Occupation_Self-employed
9. Occupation_children
10. LivesIn
11. GlucoseLevel
12. BMI
13. SmokingStatus_formerly smoked
14. SmokingStatus_never smoked
15. SmokingStatus_smokes

NOTE:
1. Make sure to preprocess the features in the above order exactly. Answer(s) of later question(s) depend(s) upon correct order of featuring processing.
2. You may have to use multiple instances of a trasnformer for this question.

In [33]:
import numpy as np
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OrdinalEncoder, OneHotEncoder, StandardScaler, MinMaxScaler

# Make a copy of X_train and X_test to preprocess
X_train_processed = X_train.copy()
X_test_processed = X_test.copy()

# Step 1: Replace implausible values with np.nan in both train and test sets
# Gender: 'Unknown' (as identified in Q2.1)
X_train_processed['Gender'] = X_train_processed['Gender'].replace('Unknown', np.nan)
X_test_processed['Gender'] = X_test_processed['Gender'].replace('Unknown', np.nan)

# Age: <= 0 (as identified in Q2.2)
X_train_processed['Age'] = X_train_processed['Age'].apply(lambda x: np.nan if x <= 0 else x)
X_test_processed['Age'] = X_test_processed['Age'].apply(lambda x: np.nan if x <= 0 else x)

# GlucoseLevel: <= 0 (as identified in Q2.3)
X_train_processed['GlucoseLevel'] = X_train_processed['GlucoseLevel'].apply(lambda x: np.nan if x <= 0 else x)
X_test_processed['GlucoseLevel'] = X_test_processed['GlucoseLevel'].apply(lambda x: np.nan if x <= 0 else x)

# LivesIn: 'Unknown' (as identified in Q2.4)
X_train_processed['LivesIn'] = X_train_processed['LivesIn'].replace('Unknown', np.nan)
X_test_processed['LivesIn'] = X_test_processed['LivesIn'].replace('Unknown', np.nan)

# SmokingStatus: 'Unknown' (as identified in Q2.6)
X_train_processed['SmokingStatus'] = X_train_processed['SmokingStatus'].replace('Unknown', np.nan)
X_test_processed['SmokingStatus'] = X_test_processed['SmokingStatus'].replace('Unknown', np.nan)

# Define preprocessing pipelines for different column types

# 1. Gender: Impute most frequent, then Ordinal Encode
# Assuming default alphabetical order for Male/Female from OrdinalEncoder is acceptable as no specific mapping is given.
gender_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('ordinal', OrdinalEncoder(handle_unknown='use_encoded_value', unknown_value=-1))
])

# 2. Age: Impute mean, then Standard Scale
age_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),
    ('scaler', StandardScaler())
])

# 3. HasTension: Ordinal Encode (No=0, Yes=1)
tension_pipeline = Pipeline(steps=[
    ('ordinal', OrdinalEncoder(categories=[['No', 'Yes']], handle_unknown='use_encoded_value', unknown_value=-1))
])

# 4. AnyHeartDisease: Ordinal Encode (No=0, Yes=1)
heartdisease_pipeline = Pipeline(steps=[
    ('ordinal', OrdinalEncoder(categories=[['No', 'Yes']], handle_unknown='use_encoded_value', unknown_value=-1))
])

# 5. NeverMarried: Ordinal Encode (No=0, Yes=1)
nevermarried_pipeline = Pipeline(steps=[
    ('ordinal', OrdinalEncoder(categories=[['No', 'Yes']], handle_unknown='use_encoded_value', unknown_value=-1))
])

# 6. Occupation: One-Hot Encode
# OneHotEncoder will sort categories alphabetically by default, matching the hint's specified order.
occupation_pipeline = Pipeline(steps=[
    ('onehot', OneHotEncoder(handle_unknown='ignore', sparse_output=False))
])

# 7. LivesIn: Impute most frequent, then Ordinal Encode (City=0, Village=1)
livesin_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('ordinal', OrdinalEncoder(categories=[['City', 'Village']], handle_unknown='use_encoded_value', unknown_value=-1))
])

# 8. GlucoseLevel: Impute mean, then Min-Max Scale
glucose_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),
    ('scaler', MinMaxScaler())
])

# 9. BMI: Impute mean, then Standard Scale
bmi_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),
    ('scaler', StandardScaler())
])

# 10. SmokingStatus: Impute most frequent, then One-Hot Encode
# OneHotEncoder will sort categories alphabetically by default, matching the hint's specified order.
smokingstatus_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore', sparse_output=False))
])

# ColumnTransformer to apply all transformations in the specified order
preprocessor = ColumnTransformer(
    transformers=[
        ('gender_proc', gender_pipeline, ['Gender']),
        ('age_proc', age_pipeline, ['Age']),
        ('tension_proc', tension_pipeline, ['HasTension']),
        ('heartdisease_proc', heartdisease_pipeline, ['AnyHeartDisease']),
        ('nevermarried_proc', nevermarried_pipeline, ['NeverMarried']),
        ('occupation_proc', occupation_pipeline, ['Occupation']),
        ('livesin_proc', livesin_pipeline, ['LivesIn']),
        ('glucose_proc', glucose_pipeline, ['GlucoseLevel']),
        ('bmi_proc', bmi_pipeline, ['BMI']),
        ('smoking_proc', smokingstatus_pipeline, ['SmokingStatus'])
    ],
    remainder='drop', # Drop any columns not explicitly transformed
    verbose_feature_names_out=False # Ensures simpler column names for transformed output
)

# Fit the preprocessor on the training data and transform both training and test data
X_train_transformed = preprocessor.fit_transform(X_train_processed)
X_test_transformed = preprocessor.transform(X_test_processed)

# Print shapes to verify
print(f"Shape of X_train_transformed: {X_train_transformed.shape}")
print(f"Shape of X_test_transformed: {X_test_transformed.shape}")

# Verify feature names (for debugging/understanding, not strictly required by the prompt but good practice)
# This helps confirm the output order matches the hint.
# feature_names_out = preprocessor.get_feature_names_out()
# print("\nTransformed feature names (order check):\n", feature_names_out)

Shape of X_train_transformed: (2800, 16)
Shape of X_test_transformed: (1200, 16)


## Q.8 [MCQ][2 Mark] Calculate the shape of the feature matrix of training dataset.

(a) (2800, 16)

(b) (2000, 16)

(c) (3000, 16)

(d) (3200, 16)

Answer:

V1: A

V2: A

V3: A




In [34]:
print(f"Shape of X_train_transformed: {X_train_transformed.shape}")

Shape of X_train_transformed: (2800, 16)


## Q.9 [NAT][6 marks] What is the mean of the transformed test data (features only)?

Note : Compute the mean of the whole feature matrix i.e. mean of all values in the transformed test feature matrix

Ans:

V1: 0.2462 [Range: 0.241 to 0.251]

V2: 0.2506 [Range: 0.245 to 0.255]

V3: 0.3870 [Range: 0.382 to 0.392]


In [35]:
import numpy as np

mean_test_data = np.mean(X_test_transformed)
print(f"Mean of the transformed test data: {mean_test_data:.4f}")

Mean of the transformed test data: 0.2462


## Q.10 [NAT][6 marks] If you eliminate 1 feature with recursive feature elimination, which feature will be eliminated?
- Type the index of the eliminated feature (index starts from 0).
- Use `LogisticRegression` model with random state as 1729 and rest of the  parameters with default values, as an estimator.
-  Use processed training data.

V1: 15

V2: 6

V3: 6

No range needed for this question.

In [36]:
from sklearn.feature_selection import RFE
from sklearn.linear_model import LogisticRegression

# Instantiate Logistic Regression model with random state 1729 and default parameters
estimator = LogisticRegression(random_state=1729, solver='liblinear') # 'liblinear' is often a good choice for default parameters

# Instantiate RFE with the estimator and specify to eliminate 1 feature.
# If n_features_to_select is not specified, it selects half the features.
# To find the 'first eliminated' feature, we can either set n_features_to_select = total_features - 1
# and then find the feature with rank 1, or set n_features_to_select=1 and find the feature with highest rank (meaning it was eliminated first).
# Let's set n_features_to_select to total_features - 1, and find the feature that was ranked as '1' for elimination.

# Number of features in the transformed training data
num_features = X_train_transformed.shape[1]

rfe_selector = RFE(estimator=estimator, n_features_to_select=num_features - 1, step=1, verbose=0)
rfe_selector.fit(X_train_transformed, y_train)

# The 'ranking_' attribute gives the rank of each feature (1 and up).
# Features not selected are ranked according to their elimination order (rank 1 is the first to be eliminated).
# The question asks for the feature that will be eliminated (singular), implying the *first* one eliminated.
# In RFE's ranking, '1' means it was the first one to be eliminated when selecting all but one.

eliminated_feature_index = np.where(rfe_selector.ranking_ == 1)[0][0]

print(f"The index of the first eliminated feature is: {eliminated_feature_index}")

The index of the first eliminated feature is: 0


# Model building
[click here to view dataset](https://drive.google.com/file/d/1HBLFUwtVdt63cFHDe9LjD1EDDEcX8jfy/view?usp=sharing)

### **Note**: For numerical type questions, always enter the answer correct upto 3 decimal places without rounding off, unless otherwise stated.

Load the dataset.
- The last column is the target column.
- Last 30% rows of the dataset constitute test set and remaining rows form the training set.
- Do not shuffle the dataset while splitting
- You must have to use only training set to train all the estimator in questions below.
- First row of the file has column names/ids, and it has no index column.

In [37]:
df_model = pd.read_csv('ModelBuilding_V1.csv')

# The last column is the target column. Let's separate features (X_model) and target (y_model).
# Assuming the last column is 'HeartAttack' based on previous context, but will confirm with .iloc
X_model = df_model.iloc[:, :-1]
y_model = df_model.iloc[:, -1]

# Replace 'Yes' with 1 and 'No' with 0 in the target variable
y_model = y_model.map({'Yes': 1, 'No': 0})

print("Dataset Head:")
print(df_model.head())

print("\nTarget column distribution:")
print(y_model.value_counts())

print("\nDataset Info:")
df_model.info()


Dataset Head:
   Gender       Age  HasTension  AnyHeartDisease  NeverMarried  \
0     0.0  0.433901         0.0              0.0           1.0   
1     1.0 -1.840435         0.0              0.0           0.0   
2     1.0 -1.160260         0.0              0.0           0.0   
3     1.0 -0.806002         0.0              0.0           1.0   
4     0.0  0.743876         0.0              0.0           1.0   

   Occupation_Govt_job  Occupation_Never_worked  Occupation_Private  \
0                  0.0                      0.0                 0.0   
1                  0.0                      0.0                 0.0   
2                  1.0                      0.0                 0.0   
3                  0.0                      0.0                 1.0   
4                  0.0                      0.0                 1.0   

   Occupation_Self-employed  Occupation_children  LivesIn  GlucoseLevel  \
0                       1.0                  0.0      1.0      0.203263   
1           

### Q.1 [MCQ][0 Mark] Which dataset are you using for this exam? Write the last two letters of the dataset file name.

A) V1

B) V2

C) V3

Answer:

V1: A

V2: B

V3: C

In [38]:
# A

### Instructions (Q2-Q3)
Instantiate a perceptron classifier that with following parameters:
  * random_state = 1729
  * learning rate = 1
  * Train for appropriate number of iterations
  * Do not shuffle the dataset for each iteration.
  * Include the intercept (bias) term.
  * Use 10% of the data for validation fraction.
  * Do not apply regularization.
  * Set warm start to true.

Hint: one iteration of training indicates going over each sample exactly once.

Train the classifier on the training data.

In [39]:
import pandas as pd
from sklearn.linear_model import Perceptron

# Ensure y_model is correctly loaded with 0/1 values from df_model.iloc[:, -1]
# The previous execution of cell 'awqyIweCzg02' resulted in y_model being all NaNs
# because the .map({'Yes': 1, 'No': 0}) was applied to an already numeric column (0s and 1s).
# We will re-assign y_model here to correct this.
y_model = df_model.iloc[:, -1]

# Manual split according to instructions: Last 30% for test, remaining for train, no shuffle.
total_rows = len(df_model)
train_size = int(total_rows * 0.7)

X_train_model = X_model.iloc[:train_size]
X_test_model = X_model.iloc[train_size:]
y_train_model = y_model.iloc[:train_size]
y_test_model = y_model.iloc[train_size:]

print(f"Shape of X_train_model: {X_train_model.shape}")
print(f"Shape of X_test_model: {X_test_model.shape}")
print(f"Shape of y_train_model: {y_train_model.shape}")
print(f"Shape of y_test_model: {y_test_model.shape}")
print(f"Distribution of y_train_model:\n{y_train_model.value_counts(normalize=True)}")
print(f"Distribution of y_test_model:\n{y_test_model.value_counts(normalize=True)}")

# Instantiate Perceptron classifier with specified parameters
perceptron_clf = Perceptron(
    random_state=1729,
    eta0=1.0,  # learning rate
    shuffle=False,  # Do not shuffle the dataset for each iteration
    fit_intercept=True,  # Include the intercept (bias) term
    validation_fraction=0.1,  # Use 10% of the data for validation fraction
    penalty=None,  # Do not apply regularization
    warm_start=True,  # Set warm start to true
    max_iter=5  # Train for 5 iterations (epochs) as specified in Q.2 instructions
)

# Train the classifier on the training data
perceptron_clf.fit(X_train_model, y_train_model)

print("\nPerceptron classifier trained successfully.")
# The trained classifier `perceptron_clf`, along with the split data, will be available in the kernel state.


Shape of X_train_model: (2800, 16)
Shape of X_test_model: (1200, 16)
Shape of y_train_model: (2800,)
Shape of y_test_model: (1200,)
Distribution of y_train_model:
HeartAttack
0    0.951429
1    0.048571
Name: proportion, dtype: float64
Distribution of y_test_model:
HeartAttack
0    0.951667
1    0.048333
Name: proportion, dtype: float64

Perceptron classifier trained successfully.


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_stochastic_gradient.py:738: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  warnings.warn(


### Q.2 [NAT] [4 Marks] Train the perceptron classifier for 5 iterations. What is value of bias (intercept) after 5th iteration?


Answer:

V1: -3.0, range: -2.995 to -3.005

V2: -4.0, range: -3.995 to -4.005

V3: -6.0, range: -5.995 to -6.005

In [40]:
print(f"Bias (intercept) after 5 iterations: {perceptron_clf.intercept_[0]:.3f}")

Bias (intercept) after 5 iterations: -3.000


###Q.3 [MCQ][5 Marks] In continuation of the previous question, compute precision accurate upto 2 decimal places on training data for positive class (i.e. class value 1), after 5 iterations.

[Hint: Use estimator trained from the previous question]


Answer:

V1: 0.666, range: 0.661 to 0.671

V2: 0.1238, range: 0.118 to 0.128

V3: 0.0, range: 0.0 to 0.005

In [41]:
from sklearn.metrics import precision_score

# Make predictions on the training data
y_train_pred = perceptron_clf.predict(X_train_model)

# Calculate precision for the positive class (class 1)
# We need to handle cases where there are no positive predictions or no actual positive samples
# 'pos_label=1' explicitly sets the positive class
precision = precision_score(y_train_model, y_train_pred, pos_label=1, zero_division=0)

print(f"Precision for class 1 on training data: {precision:.2f}")

Precision for class 1 on training data: 0.67


### Q.4 [NAT][5 Marks] Train (on training data only) logistic regression using `SGDClassifier`. Use the following parameters:

1. Choose appropriate loss value to obtain logistic regression
2. penalty='l2',
3. eta0=0.001,
4. alpha=0,
5. learning_rate='constant'
6. random_state=1729.
7. warm_start = True

Train the classifier for 5 iterations and note the value of the loss in each iteration. What will be the loss value after second iteration? Answer upto three decimal places.

**Note:** Set the remaining parameters, if any, accordingly to be able to get the loss value after second iteration. Also note that the classifier has to be trained for 5 iterations.

Answer:

V1: 0.2082 Range: 0.203 to 0.213

V2: 0.2096 Range: 0.204 to 0.214

V3: 0.2111 Range: 0.206 to 0.216

In [42]:
from sklearn.linear_model import SGDClassifier
from sklearn.metrics import log_loss
import numpy as np

# Instantiate SGDClassifier for logistic regression with specified parameters
# max_iter=1 and warm_start=True allows us to fit one epoch at a time
# tol=None ensures the training runs for max_iter regardless of convergence
sgd_clf = SGDClassifier(
    loss='log_loss',  # Appropriate loss for logistic regression
    penalty='l2',
    eta0=0.001,
    alpha=0.0, # alpha=0.0 effectively disables L2 regularization in this context
    learning_rate='constant',
    random_state=1729,
    warm_start=True,
    max_iter=1, # Fit for one iteration (epoch) at a time
    tol=None    # Disable stopping criteria to ensure 5 iterations are run
)

losses = []

# Train the classifier for 5 iterations (epochs)
for i in range(5):
    # Fit for one epoch incrementally
    sgd_clf.fit(X_train_model, y_train_model)

    # Calculate log loss on the training data after each iteration
    # predict_proba is available for 'log_loss'
    y_train_proba = sgd_clf.predict_proba(X_train_model)[:, 1]
    current_loss = log_loss(y_train_model, y_train_proba)
    losses.append(current_loss)

print(f"Loss values after each iteration: {losses}")

# The question asks for the loss value after the second iteration (index 1 in a 0-indexed list)
loss_after_second_iteration = losses[1]
print(f"Loss value after the second iteration: {loss_after_second_iteration:.3f}")

Loss values after each iteration: [0.2529904609012919, 0.20828682141739835, 0.19406901833322654, 0.18699850891012404, 0.18255077295024025]
Loss value after the second iteration: 0.208


### Q.5 [MCQ][5 marks] Use GridSearchCV with SGDClassifier. Following are the classifier's parameters:

- `loss = 'log_loss'`
- `learning_rate = 'constant'`
- `random_state = 1729`

Following are parameters to examine:
-  `alpha = [0.0001, 0.0005, 0.001, 0.005]`
- `eta0 = [0.01, 0.05, 0.1, 0.5]`

What are the best values of alpha and eta0 respectively?

A. 0.0001, 0.01

B. 0.001, 0.01

C. 0.0001, 0.5

D. 0.0005, 0.01

E. 0.005, 0.5

F 0.05, 0.1

Answer:

V1: A

V2: A

V3: F

In [43]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import SGDClassifier

# Instantiate SGDClassifier with fixed parameters
sgd_clf_grid = SGDClassifier(
    loss='log_loss',
    learning_rate='constant',
    random_state=1729,
    warm_start=True, # warm_start is useful for iterative models, but not strictly needed for GridSearchCV's single fit per combination
    max_iter=1000,   # Set a reasonable max_iter for convergence for GridSearchCV
    tol=1e-3         # Set tolerance for stopping criteria
)

# Define the parameter grid to search
param_grid = {
    'alpha': [0.0001, 0.0005, 0.001, 0.005],
    'eta0': [0.01, 0.05, 0.1, 0.5]
}

# Instantiate GridSearchCV
# Using 'accuracy' as a scoring metric, you might want to consider others for imbalanced datasets
grid_search = GridSearchCV(estimator=sgd_clf_grid, param_grid=param_grid, cv=5, scoring='accuracy', verbose=1, n_jobs=-1)

# Fit GridSearchCV on the training data
grid_search.fit(X_train_model, y_train_model)

print(f"Best parameters found: {grid_search.best_params_}")
print(f"Best accuracy score: {grid_search.best_score_:.4f}")

Fitting 5 folds for each of 16 candidates, totalling 80 fits
Best parameters found: {'alpha': 0.0001, 'eta0': 0.01}
Best accuracy score: 0.9514


### Q.6 [MCQ][5 Marks] Create a new estimator using SGDClassifier that uses the best parameters obtained in Gridsearch earlier ((learning rate to be constant, random_state to be '1729' and use appropriate loss for logistic regression)) and set the weight of class 0 to be 0.1 and the weight of class 1 to be 2. How many samples of class 1 from the test set are correctly predicted by this estimator?



A. 47

B. 46

C. 55

D. 40

E. 41

F. None of these

Ans:

V1: A

V2: B

V3: E

In [44]:
from sklearn.linear_model import SGDClassifier
from sklearn.metrics import confusion_matrix

# Retrieve best parameters from the GridSearchCV object
# Assuming grid_search object is available from the previous cell's execution
best_alpha = grid_search.best_params_['alpha']
best_eta0 = grid_search.best_params_['eta0']

# Create a new SGDClassifier with best parameters and specified class weights
sgd_clf_weighted = SGDClassifier(
    loss='log_loss',  # For logistic regression
    learning_rate='constant',
    random_state=1729,
    penalty='l2', # L2 is default for SGDClassifier and not explicitly set to 0 in this question
    eta0=best_eta0,
    alpha=best_alpha,
    class_weight={0: 0.1, 1: 2}, # Set custom class weights
    max_iter=1000, # A reasonable number of iterations for convergence
    tol=1e-3 # Tolerance for stopping criterion
)

# Train the classifier on the training data
sgd_clf_weighted.fit(X_train_model, y_train_model)

# Make predictions on the test data
y_test_pred_weighted = sgd_clf_weighted.predict(X_test_model)

# Filter for actual class 1 samples in the test set
actual_class_1_indices = y_test_model[y_test_model == 1].index

# Get predictions for these actual class 1 samples
predictions_for_actual_class_1 = y_test_pred_weighted[actual_class_1_indices - len(X_train_model)] # Adjust index for test_pred array

# Count correctly predicted class 1 samples
correctly_predicted_class_1 = (predictions_for_actual_class_1 == 1).sum()

print(f"Number of samples of class 1 from the test set correctly predicted: {correctly_predicted_class_1}")

# For verification, you can also print the confusion matrix:
# cm = confusion_matrix(y_test_model, y_test_pred_weighted)
# print("\nConfusion Matrix on Test Data:\n", cm)
# print(f"True Positives (correctly predicted class 1): {cm[1, 1]}")

Number of samples of class 1 from the test set correctly predicted: 47


###Q.7 [MCQ][ 5 Marks] Fit an SVM classifier with following parameters:

- `kernel='rbf'`
- `decision_function_shape='ovr'`
- random_state=`1729`  
- `C=1`

Train the model on training data, and print the confusion matrix on test data.

(MCQ)

Options:

(a) [[1142    0]
 [  58    0]]

 (b)  [[1141    0]
 [  59    0]]


 (c)  [[1139    0]
 [  61    0]]

 (d) [[1150   0]
 [ 50   0]]

 (e) [[1000    0]
 [ 200    0]]


Answer

V1: A

V2: B

V3: C

In [45]:
from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix

# Instantiate SVM classifier with specified parameters
svm_clf = SVC(
    kernel='rbf',
    decision_function_shape='ovr',
    random_state=1729,
    C=1
)

# Train the model on training data
svm_clf.fit(X_train_model, y_train_model)

# Make predictions on the test data
y_test_pred_svm = svm_clf.predict(X_test_model)

# Compute and print the confusion matrix on test data
cm_svm = confusion_matrix(y_test_model, y_test_pred_svm)
print("Confusion Matrix on Test Data:")
print(cm_svm)

Confusion Matrix on Test Data:
[[1142    0]
 [  58    0]]


###Instructions for Q8-10
Train a Decision Tree Classifier with the following properties:

- criterion = 'entropy',
- splitter = 'random',
- min_samples_split = 4,
- min_impurity_decrease = 0.0001,
- random_state = 1729



In [46]:
from sklearn.tree import DecisionTreeClassifier

# Instantiate Decision Tree Classifier with specified parameters
dtc_clf = DecisionTreeClassifier(
    criterion='entropy',
    splitter='random',
    min_samples_split=4,
    min_impurity_decrease=0.0001,
    random_state=1729
)

# Train the model on training data
dtc_clf.fit(X_train_model, y_train_model)

print("Decision Tree Classifier trained successfully.")
# The trained classifier `dtc_clf` will be available in the kernel state for subsequent questions.

Decision Tree Classifier trained successfully.


###Q.8 [MCQ][5 Marks] What is the resultant depth of the tree?


A. 18

B. 20

C. 21

D. 24

E. 26

Answer:

V1: B

V2: B

V3: A

In [47]:
print(f"Resultant depth of the tree: {dtc_clf.tree_.max_depth}")

Resultant depth of the tree: 20


### Q.9 [MCQ][Marks 5] How many nodes are there in the tree?


A. 495

B. 515

C. 491

D. 589

E. 571

Answer:

V1: B (515)

V2: A (495)

V3: C (491)

In [48]:
print(f"Number of nodes in the tree: {dtc_clf.tree_.node_count}")

Number of nodes in the tree: 515


###Q.10 [NAT][5 Marks] What is the value of entropy at the left child of root?

Answer:

V1: 0.025 Range:0.0245 to 0.0255

V2: 0.032 Range: 0.027 to 0.037

V3: 0.025 Range:0.0245 to 0.0255

In [49]:
import numpy as np

# The root node is typically index 0.
# The left child of the root is typically index 1.
# The impurity (entropy in this case, as criterion='entropy' was used) for each node
# is stored in dtc_clf.tree_.impurity.

# Get the entropy of the left child of the root node
# dtc_clf.tree_.children_left[0] gives the index of the left child of the root (node 0).
left_child_root_index = dtc_clf.tree_.children_left[0]

# Get the impurity (entropy) of that node
entropy_left_child = dtc_clf.tree_.impurity[left_child_root_index]

print(f"Entropy at the left child of root: {entropy_left_child:.3f}")

Entropy at the left child of root: 0.025


###Q. 11 [MCQ][8 Marks] Out of DecisionTreeClassifier, KNeighborsClassifier and LogisticRegression, which one performs the best when used as base estimator in BaggingClassifier on the test data in terms of accuracy score when 20 base estimators are used ?

(Use random state 1729 for BaggingClassifier, DecisionTreeClassifier and LogisticRegression)

The metric for best performance will be the lowest 'absolute' difference in the train and test score.

A. DecisionTreeClassifier

B. KNeighborsClassifier

C. LogisticRegression

Answer:

V1: C

V2: C

V3: C



In [50]:
import pandas as pd
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# --- Re-establish data splits (as y_model in kernel state is NaN) ---
# Assuming df_model is loaded from 'ModelBuilding_V1.csv' in a previous cell
# If df_model is not in the kernel state, this cell would need to load it.
# For safety, let's assume df_model was loaded correctly.

# The last column is the target column. Let's separate features (X_model) and target (y_model).
# Re-loading df_model and splitting to ensure correct y_model values
# This code block needs to be self-contained for the variable definitions.
df_model = pd.read_csv('ModelBuilding_V1.csv')
X_model = df_model.iloc[:, :-1]
y_model = df_model.iloc[:, -1]

# The following line caused the error because 'HeartAttack' is already 0/1 integers.
# y_model = y_model.map({'Yes': 1, 'No': 0})
# Removing this line as the data is already numerical.

# Manual split according to instructions: Last 30% for test, remaining for train, no shuffle.
total_rows = len(df_model)
train_size = int(total_rows * 0.7)

X_train_model = X_model.iloc[:train_size]
X_test_model = X_model.iloc[train_size:]
y_train_model = y_model.iloc[:train_size]
y_test_model = y_model.iloc[train_size:]
# ----------------------------------------------------------------------

results = {}
estimators = {
    "DecisionTreeClassifier": DecisionTreeClassifier(random_state=1729),
    "KNeighborsClassifier": KNeighborsClassifier(),
    "LogisticRegression": LogisticRegression(random_state=1729, solver='liblinear') # Using liblinear for consistency
}

for name, estimator in estimators.items():
    print(f"\n--- Evaluating {name} as base estimator ---")

    bagging_clf = BaggingClassifier(
        estimator=estimator,
        n_estimators=20,
        random_state=1729,
        n_jobs=-1 # Use all available cores for faster training
    )

    # Train the BaggingClassifier
    bagging_clf.fit(X_train_model, y_train_model)

    # Predict on training data
    y_train_pred = bagging_clf.predict(X_train_model)
    train_accuracy = accuracy_score(y_train_model, y_train_pred)

    # Predict on test data
    y_test_pred = bagging_clf.predict(X_test_model)
    test_accuracy = accuracy_score(y_test_model, y_test_pred)

    # Calculate absolute difference
    abs_diff = abs(train_accuracy - test_accuracy)

    results[name] = {"train_accuracy": train_accuracy, "test_accuracy": test_accuracy, "abs_diff": abs_diff}

    print(f"  Train Accuracy: {train_accuracy:.4f}")
    print(f"  Test Accuracy: {test_accuracy:.4f}")
    print(f"  Absolute Difference (Train-Test): {abs_diff:.4f}")

# Determine the best performing estimator
best_estimator = min(results, key=lambda k: results[k]['abs_diff'])
lowest_abs_diff = results[best_estimator]['abs_diff']

print(f"\n--- Conclusion ---")
print(f"The best performing estimator based on the lowest absolute difference (train-test score) is: {best_estimator}")
print(f"Lowest Absolute Difference: {lowest_abs_diff:.4f}")



--- Evaluating DecisionTreeClassifier as base estimator ---
  Train Accuracy: 0.9939
  Test Accuracy: 0.9442
  Absolute Difference (Train-Test): 0.0498

--- Evaluating KNeighborsClassifier as base estimator ---
  Train Accuracy: 0.9521
  Test Accuracy: 0.9508
  Absolute Difference (Train-Test): 0.0013

--- Evaluating LogisticRegression as base estimator ---
  Train Accuracy: 0.9514
  Test Accuracy: 0.9517
  Absolute Difference (Train-Test): 0.0002

--- Conclusion ---
The best performing estimator based on the lowest absolute difference (train-test score) is: LogisticRegression
Lowest Absolute Difference: 0.0002


### Q. 12 [NAT] [8 Marks] When the above three individual classifiers (with same settings) are used in VotingClassifier, how much absolute difference do we obtain in train and test scores? Enter your answer correct upto 4 decimal places.

Answer:

V1: 0.0042 Range : 0.0037 to 0.0047

V2: 0.00654 range: 0.006 to 0.007

V3: 0.00821 Range: 0.0077 0.0087

In [51]:
from sklearn.ensemble import VotingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
import numpy as np

# --- Re-establish data splits (if not already in kernel state, though it should be from Q11) ---
# This block is for robustness; ideally, X_train_model, y_train_model, etc. are already defined.
if 'X_train_model' not in locals():
    df_model = pd.read_csv('ModelBuilding_V1.csv')
    X_model = df_model.iloc[:, :-1]
    y_model = df_model.iloc[:, -1]
    y_model = y_model.map({'Yes': 1, 'No': 0})
    total_rows = len(df_model)
    train_size = int(total_rows * 0.7)
    X_train_model = X_model.iloc[:train_size]
    X_test_model = X_model.iloc[train_size:]
    y_train_model = y_model.iloc[:train_size]
    y_test_model = y_model.iloc[train_size:]
# -------------------------------------------------------------------------------------------------

# Instantiate the three individual classifiers with the specified settings
est1 = DecisionTreeClassifier(random_state=1729)
est2 = KNeighborsClassifier()
est3 = LogisticRegression(random_state=1729, solver='liblinear') # Using liblinear for consistency

# Create a VotingClassifier
voting_clf = VotingClassifier(estimators=[
    ('dt', est1),
    ('knn', est2),
    ('lr', est3)
],
    voting='hard', # 'hard' voting for classifiers
    n_jobs=-1 # Use all available cores
)

# Train the VotingClassifier on the training data
voting_clf.fit(X_train_model, y_train_model)

# Calculate accuracy on the training data
y_train_pred_voting = voting_clf.predict(X_train_model)
train_accuracy_voting = accuracy_score(y_train_model, y_train_pred_voting)

# Calculate accuracy on the test data
y_test_pred_voting = voting_clf.predict(X_test_model)
test_accuracy_voting = accuracy_score(y_test_model, y_test_pred_voting)

# Calculate the absolute difference
absolute_difference = abs(train_accuracy_voting - test_accuracy_voting)

print(f"Train Accuracy (VotingClassifier): {train_accuracy_voting:.4f}")
print(f"Test Accuracy (VotingClassifier): {test_accuracy_voting:.4f}")
print(f"Absolute difference in train and test scores: {absolute_difference:.4f}")

Train Accuracy (VotingClassifier): 0.9543
Test Accuracy (VotingClassifier): 0.9500
Absolute difference in train and test scores: 0.0043
